sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/QueryPlanner.scala

```diff
@@ -81,7 +81,7 @@ abstract class QueryPlanner[PhysicalPlan <: TreeNode[PhysicalPlan]] {
               childPlans.map { childPlan =>
                 // Replace the placeholder by the child plan
                 candidateWithPlaceholders.transformUp {
-                  case p if p == placeholder => childPlan
+                  case p if p.eq(placeholder) => childPlan
```

Comment:
good catch... this fix is simple and ok to me in this case though, I think we'd be better to compare placeholder nodes only here, right?

Reply:
not sure what you mean exactly here, may you elaborate a bit more please? Thanks.

Reply:
nvm, just a quesion. I was thinking why we couldn't write here like;
```
trait PlaceHolder;
case class PlanLater extends LeafExecNode with PlaceHolder;
```
then,
```
candidateWithPlaceholders.transformUp {
  case p: PlaceHolder if p.eq(placeholder) => childPlan
}
```

Reply:
I think we could do that only in 3.0, as it would be a breaking change for those who are using a custom QueryPlanner.

Reply:
yea, thanks. anyway, this fix looks pretty ok.

Reply:
Probably we can also move `PlanLater` to catalyst and use that instead of introducing a new trait. I think it can be proposed for 3.0. Thanks.